crates/deltalake-core/src/protocol/checkpoints.rs

```diff
@@ -117,6 +122,14 @@ pub async fn create_checkpoint_for(
     state: &DeltaTableState,
     log_store: &dyn LogStore,
 ) -> Result<(), ProtocolError> {
+    if version != state.version() {
+        error!(
+            "create_checkpoint_for called with version {version} but table state contains: {}. The table state may need to be reloaded",
+            state.version()
+        );
+        return Err(CheckpointError::StaleTableVersion(version, state.version()).into());
```

Comment:
Why can't we just checkout the correct version?

Reply:
@wjones127 how do you mean "checkout?" I am less versed in the nuance of checkpoints. If the last version in the table is 100, is it valid to call this with `_last_checkpoint` of `version: 5`?

I honestly don't know whether we should even allow a user-specified version here at all, rather than creating a checkpoint with the version of the table state.

HOWEVER, if a writer has a stale table state creates `_last_checkpoint` after _another_ writer has created a later checkpoint, this becomes confusing and out of sync anyways :confounded:

Reply:
I guess to expand, I like the failure option here because the caller is asking us to do something here that will result in incorrect state, and we don't know what they intend or what performance concerns they have. I.e. I don't think we should reload the state here, as an example, because that would be surprising and may actually result in a later version than the user specified too

Reply:
I feel like (a) failing is the easiest solution that avoids any potential new bug introduction and (b) reloading correct version could have unexpected perf impact, esp for things like streaming workloads that are managing their own table state

Reply:
oh one more thing since my thoughts are disjointed here :laughing: I think we already have the path you describe covered with `create_checkpoint()`  :smile:

Reply:
Okay re-read the signature. Erroring makes sense to me.

I was thinking in the situation you were describing in the issue report, it sounds like the correct thing to do is to checkout the desired version first then create the checkpoint.